sirepo/sim_db_file.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+"""Getting and putting simulation db files
+
+:copyright: Copyright (c) 2020 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+from pykern import pkio
+from pykern.pkdebug import pkdp
+from pykern.pkcollections import PKDict
+import functools
+import sirepo.job
+import sirepo.simulation_db
+import sirepo.tornado
+import tornado.web
+
+_TOKEN_TO_UID = PKDict()
+
+
+class FileReq(tornado.web.RequestHandler):
+
+    def delete(self, path):
+        self.__validate_req()
+        for f in pkio.sorted_glob(sirepo.srdb.root().join(path + '*')):
+            pkio.unchecked_remove(f)
+
+    def get(self, path):
+        self.__validate_req()
+        p = sirepo.srdb.root().join(path)
+        if not p.exists():
+            raise sirepo.tornado.error_not_found()
+        self.write(pkio.read_binary(p))
+
+    def put(self, path):
+        self.__validate_req()
+        sirepo.srdb.root().join(path).write_binary(self.request.body)
+
+    def __validate_req(self):
+        t = self.request.headers.get(sirepo.job.AUTH_HEADER)
+        if not t:
+            raise sirepo.tornado.error_forbidden()
+        p = t.split(' ')
+        if len(p) != 2:
+            raise sirepo.tornado.error_forbidden()
+        if p[0] != sirepo.job.AUTH_HEADER_SCHEME_BEARER \
+            or not sirepo.job.UNIQUE_KEY_RE.search(p[1]):
+            raise sirepo.tornado.error_forbidden()
+        if p[1] not in _TOKEN_TO_UID:
+            raise sirepo.tornado.error_forbidden()
+        sirepo.simulation_db.validate_path(
+            self.request.path.split('/')[2:],
+            _TOKEN_TO_UID[p[1]],
+        )
+
+
+def get_token(uid):
```

Comment:
The code is only called by the driver and the driver only runs in the supervisor. Is that enough of a guarantee? I could add a POSIT?

Reply:
Not enough. It's a critical section. I wonder what the equivalent of `async def` is for something like this. Do you know?

Reply:
Can you elaborate a bit more? I don't know what you mean by "equivalent of `async def`".

Reply:
I think I mean `async with`. However, then we would need to define a Lock. I think this might be worthwhile, actually, since we may be using `with _global_lock` in simulation_db. There's an assumption in simulation_db that those calls are not happening in the supervisor. I think we should change that to an assertion. If running in tornado, `with _global_lock` should throw an exception. In sim_db_file, we should throw an exception if called from flask. This implies that flask and tornado distinction should be known at program startup, which it can be. The supervisor and server would call something like sirepo.util.init, which would initialize either sirepo.util.SIMULATION_DB_LOCK (replacing _global_lock) or SIM_DB_FILE_LOCK (to protect TOKEN_TO_UID). The uninitialized values would cause `with` to die, because SIM_DB_FILE_LOCK would be None in the server and SIMULATION_DB_LOCK would be None in the supervisor.

Reply:
Thanks for the details! 893b825f

The "lock" for sim_db_file isn't actually a lock. It is just a null context manager. I went this route because:
1. It doesn't actually need to be a lock. There are no awaits in the code.
2. If it is a lock then we will have to change where we call `token_for_user`. It can't be called in `__init__` because it will need an `await`.

Reply:
OK, makes sense. That's actually a real problem in tornado init (and with being explicit with the async abstractions).